File: bindings/python/cas.py
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Dict, List


class IdentityType(str, Enum):
    AGENT = "agent"
    MODEL = "model"
    OPERATOR = "operator"


@dataclass
class Identity:
    id: str
    type: IdentityType
    metadata: Dict[str, Any] = field(default_factory=dict)


@dataclass
class Span:
    id: str
    run_id: str
    type: str
    timestamp: int
    data: Dict[str, Any] = field(default_factory=dict)


@dataclass
class Receipt:
    run_id: str
    hash: str
    spans: List[Span]
    result: Any
    created_at: str

File: bindings/rust/src/cas.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Identity {
    pub id: String,
    pub r#type: IdentityType,
    #[serde(default)]
    pub metadata: HashMap<String, serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum IdentityType {
    Agent,
    Model,
    Operator,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Run {
    pub run_id: String,
    pub identity: Identity,
    pub payload: serde_json::Value,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Span {
    pub id: String,
    pub run_id: String,
    pub r#type: String,
    pub timestamp: i64,
    #[serde(default)]
    pub data: HashMap<String, serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Receipt {
    pub run_id: String,
    pub hash: String,
    pub spans: Vec<Span>,
    pub result: serde_json::Value,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Fault {
    pub run_id: String,
    pub invariant_id: String,
    pub message: String,
    pub timestamp: String,
}
